crates/nu-command/src/network/http/client.rs

```diff
@@ -200,19 +212,57 @@ pub fn send_request(
                 })
                 .collect::<Result<Vec<(String, String)>, ShellErrorOrRequestError>>()?;
 
-            let data = data
-                .iter()
-                .map(|(a, b)| (a.as_str(), b.as_str()))
-                .collect::<Vec<(&str, &str)>>();
-
-            request.send_form(&data).map_err(error_handler)
+            let request_fn = move || {
+                // coerce `data` into a shape that send_form() is happy with
+                let data = data
+                    .iter()
+                    .map(|(a, b)| (a.as_str(), b.as_str()))
+                    .collect::<Vec<(&str, &str)>>();
+                request.send_form(&data)
+            };
+            send_cancellable_request(&request_url, Box::new(request_fn), ctrl_c)
         }
         _ => Err(ShellErrorOrRequestError::ShellError(ShellError::IOError(
             "unsupported body input".into(),
         ))),
     }
 }
 
+// Helper method used to make blocking HTTP request calls cancellable with ctrl+c
+// ureq functions can block for a long time (default 30s?) while attempting to make an HTTP connection
+fn send_cancellable_request(
+    request_url: &str,
+    request_fn: Box<dyn FnOnce() -> Result<Response, Error> + Sync + Send>,
+    ctrl_c: Option<Arc<AtomicBool>>,
+) -> Result<Response, ShellErrorOrRequestError> {
+    let (tx, rx) = mpsc::channel::<Result<Response, Error>>();
+
+    // Make the blocking request on a background thread...
+    std::thread::Builder::new()
+        .name("HTTP requester".to_string())
+        .spawn(move || {
+            let ret = request_fn();
+            let _ = tx.send(ret); // may fail if the user has cancelled the operation
+        })
+        .expect("Failed to create thread");
+
+    // ...and poll the channel for responses
+    loop {
+        if nu_utils::ctrl_c::was_pressed(&ctrl_c) {
+            // Return early and give up on the background thread. The connection will either time out or be disconnected
+            return Err(ShellErrorOrRequestError::ShellError(
+                ShellError::InterruptedByUser { span: None },
+            ));
+        }
+
+        if let Ok(result) = rx.try_recv() {
+            return result
+                .map_err(|e| ShellErrorOrRequestError::RequestError(request_url.to_string(), e));
+        }
+        std::thread::sleep(Duration::from_millis(100)); // arbitrarily chosen interval
```

Comment:
While this prevents this from turning into a burning spin lock I am not sure if there is the potential for this to stall other stuff.

I think the more appropriate alternative is to use https://doc.rust-lang.org/std/sync/mpsc/struct.Receiver.html#method.recv_timeout

Reply:
Good call, done!

Reply:
We still need to continue the loop when reaching the channel time out.

Reply:
Sorry misread the location of `.map_err` on the mobile diff. I personally would be defensive and panic or "nushell failed" when receiving another error than time out.